### ISSUE TITLE
[JENKINS-44958] File handle leaks in *.log after hard kill

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/workflow-cps-plugin#1327]
<p>I use a Jenkins Pipeline to run my individual tests in parallel. I just noticed my test job that I run in parallel is failing with "Too many open files" errors.</p>

<p>Next, I ssh into the master and did an lsof -p using the pid of Jenkins... i see over 4k file descriptors. The output has thousands of lines that look like this:</p>

<pre>
java 7 root 4089w REG 202,96 28 2375888 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15331/222.log
java 7 root 4090w REG 202,96 328 787837 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15343/166.log
java 7 root 4091w REG 202,96 328 787851 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15343/180.log
java 7 root 4094w REG 202,96 671 787839 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15343/167.log
java 7 root 4095w REG 202,96 29 2361097 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15331/157.log</pre>
<p>Each file has content that roughly looks like this:</p>

<pre>
Starting building: ESC[8mha:<span class="code-comment">////4CQAnoPVmHzYGq9xQUbp+LIf+xF38/zSfY1nZ0iZCPexAAAArB+LCAAAAAAAAP9b85aBtbiIQTGjNKU4P08vOT+vOD8nVc83PyU1x6OyILUoJzMv2y+/JJUBAhiZGBgqihhk0NSjKDWzXb3Rdl
</span>LBUSYGJk8GtpzUvPSSDB8G5tKinBIGIZ+sxLJE/ZzEvHT94JKizLx0a6BxUmjGOUNodHsLgAyZEgZF/az8JP2U/Pyi4pLUAt2ULKBh+fElqcUl+sYWliYGhvoAbOU0btYAAAA=ESC[0mmy_test #389401</pre>
<p> </p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/piratejohnny.jpg" title="piratejohnny's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=piratejohnny">piratejohnny</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44958" target="_blank">File handle leaks in *.log after hard kill</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: workflow-cps-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2018-10-04T22:22:29+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-12-07
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I use a Jenkins Pipeline to run my individual tests in parallel. I just noticed my test job that I run in parallel is failing with "Too many open files" errors.</p>

<p>Next, I ssh into the master and did an lsof -p using the pid of Jenkins... i see over 4k file descriptors. The output has thousands of lines that look like this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">java 7 root 4089w REG 202,96 28 2375888 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15331/222.log
java 7 root 4090w REG 202,96 328 787837 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15343/166.log
java 7 root 4091w REG 202,96 328 787851 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15343/180.log
java 7 root 4094w REG 202,96 671 787839 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15343/167.log
java 7 root 4095w REG 202,96 29 2361097 /<span class="code-keyword">var</span>/jenkins_home/jobs/test_pipeline/builds/15331/157.log</pre>
</div></div>
<p>Each file has content that roughly looks like this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Starting building: ESC[8mha:<span class="code-comment">////4CQAnoPVmHzYGq9xQUbp+LIf+xF38/zSfY1nZ0iZCPexAAAArB+LCAAAAAAAAP9b85aBtbiIQTGjNKU4P08vOT+vOD8nVc83PyU1x6OyILUoJzMv2y+/JJUBAhiZGBgqihhk0NSjKDWzXb3Rdl
</span>LBUSYGJk8GtpzUvPSSDB8G5tKinBIGIZ+sxLJE/ZzEvHT94JKizLx0a6BxUmjGOUNodHsLgAyZEgZF/az8JP2U/Pyi4pLUAt2ULKBh+fElqcUl+sYWliYGhvoAbOU0btYAAAA=ESC[0mmy_test #389401</pre>
</div></div>
<p> </p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Jenkins ver. 2.46.3</code></li></ul>
<details><summary><i>3 attachments</i></summary>

- [image-2017-06-27-12-41-39-630.png](https://issues.jenkins.io/secure/attachment/38670/image-2017-06-27-12-41-39-630.png)
  > ![image-2017-06-27-12-41-39-630.png](https://issues.jenkins.io/secure/attachment/38670/image-2017-06-27-12-41-39-630.png)
- [image-2017-09-09-09-48-57-939.png](https://issues.jenkins.io/secure/attachment/39645/image-2017-09-09-09-48-57-939.png)
  > ![image-2017-09-09-09-48-57-939.png](https://issues.jenkins.io/secure/attachment/39645/image-2017-09-09-09-48-57-939.png)
- [image-2017-09-09-09-49-15-788.png](https://issues.jenkins.io/secure/attachment/39646/image-2017-09-09-09-49-15-788.png)
  > ![image-2017-09-09-09-49-15-788.png](https://issues.jenkins.io/secure/attachment/39646/image-2017-09-09-09-49-15-788.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/agupta.svg" title="agupta's avatar" /> agupta:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44958?focusedId=314468&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-314468">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> - I am running into same issues. Open file handlers are mainly log files from jobs which are no longer running. To note, I have not killed any job which could lead to this behavior as mentioned. I believe it is a bug with the way pipeline jobs are managing logs. I looked at <a href="https://issues.jenkins-ci.org/browse/JENKINS-38381" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-38381</a> but doesn't really understand what change needs to be made to fix this issue.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> - I am running into same issues. Open file handlers are mainly log files from jobs which are no longer running. To note, I have not killed any job which could lead to this behavior as mentioned. I believe it is a bug with the way pipeline jobs are managing logs. I looked at <a href="https://issues.jenkins-ci.org/browse/JENKINS-38381" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-38381</a> but doesn't really understand what change needs to be made to fix this issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/agupta.svg" title="agupta's avatar" /> agupta:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44958?focusedId=316472&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316472">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Unfortunately, I don't have exact steps to reproduce. If you have any steps which can help us identify the steps to reproduce this, that would help. </p>

<p>Also, is there a workaround to this issue?</p>
</pre>
</details>

</ul>
</details>

<p>Unfortunately, I don't have exact steps to reproduce. If you have any steps which can help us identify the steps to reproduce this, that would help. </p>

<p>Also, is there a workaround to this issue?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/piratejohnny.jpg" title="piratejohnny's avatar" /> piratejohnny:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44958?focusedId=316550&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316550">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I just use Jenkins less and I also relaunched my master on a bigger AWS instance and I haven't seen it since then.</p>

<p>I'm 99% sure that the biggest change that made it go away was launching the master on a bigger instance.</p>
</pre>
</details>

</ul>
</details>

<p>I just use Jenkins less and I also relaunched my master on a bigger AWS instance and I haven't seen it since then.</p>

<p>I'm 99% sure that the biggest change that made it go away was launching the master on a bigger instance.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>freemail165:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44958?focusedId=342832&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-342832">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This happens not only for pipeline but also for job</p>
</pre>
</details>

</ul>
</details>

<p>This happens not only for pipeline but also for job</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>freemail165:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44958?focusedId=342833&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-342833">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>looks like when builds reach the limitation, jenkins delete expired builds but JVM still keep the link to those logs, that is the issue</p>

<p>ps -ef | grep java </p>

<p>ls -tl /proc/*****/fd | grep deleted</p>

<p> </p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 912 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66892/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 913 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66889/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 914 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66893/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 916 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66997/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 917 -&gt; /var/lib/jenkins/jobs/performance-ios-build/builds/.4574/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 918 -&gt; /var/lib/jenkins/jobs/performance-android-build/builds/.14546/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 920 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66933/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 921 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66888/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 922 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66891/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 924 -&gt; /var/lib/jenkins/jobs/performance-android-build/builds/.14654/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 927 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66894/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 928 -&gt; /var/lib/jenkins/jobs/performance-android-build/builds/.14568/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 929 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66896/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 931 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66895/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 932 -&gt; /var/lib/jenkins/jobs/metrics-report/builds/.66939/log (<b>deleted</b>)</p>
</pre>
</details>

</ul>
</details>

<p>looks like when builds reach the limitation, jenkins delete expired builds but JVM still keep the link to those logs, that is the issue</p>

<p>ps -ef | grep java </p>

<p>ls -tl /proc/*****/fd | grep deleted</p>

<p> </p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 912 -> /var/lib/jenkins/jobs/metrics-report/builds/.66892/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 913 -> /var/lib/jenkins/jobs/metrics-report/builds/.66889/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 914 -> /var/lib/jenkins/jobs/metrics-report/builds/.66893/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 916 -> /var/lib/jenkins/jobs/metrics-report/builds/.66997/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 917 -> /var/lib/jenkins/jobs/performance-ios-build/builds/.4574/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 918 -> /var/lib/jenkins/jobs/performance-android-build/builds/.14546/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 920 -> /var/lib/jenkins/jobs/metrics-report/builds/.66933/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 921 -> /var/lib/jenkins/jobs/metrics-report/builds/.66888/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 922 -> /var/lib/jenkins/jobs/metrics-report/builds/.66891/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 924 -> /var/lib/jenkins/jobs/performance-android-build/builds/.14654/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 927 -> /var/lib/jenkins/jobs/metrics-report/builds/.66894/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 928 -> /var/lib/jenkins/jobs/performance-android-build/builds/.14568/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 929 -> /var/lib/jenkins/jobs/metrics-report/builds/.66896/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 931 -> /var/lib/jenkins/jobs/metrics-report/builds/.66895/log (<b>deleted</b>)</p>

<p>lr-x------ 1 jenkins jenkins 64 Jun 20 22:23 932 -> /var/lib/jenkins/jobs/metrics-report/builds/.66939/log (<b>deleted</b>)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44958?focusedId=343027&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-343027">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=freemail165" class="user-hover" rel="freemail165">freemail165</a> Are you running the latest core and plugin versions?  If so, can you provide a simple case that will reproduce the issue from scratch?</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=freemail165" class="user-hover" rel="freemail165">freemail165</a> Are you running the latest core and plugin versions?  If so, can you provide a simple case that will reproduce the issue from scratch?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>freemail165:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44958?focusedId=343306&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-343306">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>core version means jenkins version? for plugin version, which plugin do u care about?</p>
</pre>
</details>

</ul>
</details>

<p>core version means jenkins version? for plugin version, which plugin do u care about?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>svanoort:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-44958?focusedId=343341&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-343341">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=freemail165" class="user-hover" rel="freemail165">freemail165</a> Yes, "core" means Jenkins version.  For plugins, we want the workflow-job, workflow-cps, durable-task, workflow-support, workflow-api, and workflow-basic-steps.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=freemail165" class="user-hover" rel="freemail165">freemail165</a> Yes, "core" means Jenkins version.  For plugins, we want the workflow-job, workflow-cps, durable-task, workflow-support, workflow-api, and workflow-basic-steps.</p>
